### ISSUE TITLE
Diffuse Irradiance Optimization for Single Axis Trackers

### ISSUE THREAD

**Issue by kurt-rhee** [pvlib/pvlib-python#1694]
**Is your feature request related to a problem? Please describe.**
Some tracker manufacturers have implemented a diffuse irradiance optimization algorithm into their controls.  The algorithm determines which angle the tracker should track to in order to maximize plane of array irradiance given the current irradiance conditions in terms of DHI and DNI.  

**Describe the solution you'd like**
I would like to model a generic diffuse irradiance optimization algorithm in pvlib.  

**Describe alternatives you've considered**
Instead of implementing this feature inside of the library, it could be posted as a user guide.  

**Additional context**
PlantPredict is about to release a new 2 parameter diffuse irradiance optimization algorithm in our next major version release.  The two parameters are meant to reduce overestimation from irradiance optimization algorithms which are too ideal.  I used pvlib in order to test the algorithm before having it implemented in C# and would be happy to contribute the code to pvlib (if there is any interest from the maintainers/community of course).  The is no paper yet unfortunately.  My hope is to publish at next year's PVSC. 

There is some available documentation here:

https://terabase.atlassian.net/servicedesk/customer/portal/3/article/1260060675

High level notes:
- For loop through all tracking angles from minimum rotation angle to maximum rotation angle with a step of 1 degree
- Choose whichever angle gave the best plane of array irradiance
- If a tracker rotation speed or a hesitation factor are != 0
- Calculate what percentage of the timestep the tracker is in motion, what percentage of the timestep the tracker hesitates before beginning its movement towards the ideal tracker angle
- Create a corrected angle which is the weighted average angle between the ideal tracker angle, the in transit tracker angle and the hesitation tracker angle.
-  If tracker rotation speed and hesitation factor are == 0
-  Diffuse optimization algorithm is ideal
- Return the corrected angle

**Comment by echedey-ls** [comment]
I'm going to address this issue if no objections are made.

The _high level notes_ section is the methodology used in order to define the optimization algorithm, right? I still have to read a few papers before a PR. But from the web page, a PVLIB implementation seams feasible.

**Comment by kurt-rhee** [comment]
Hey @echedey-ls that is correct, if it is helpful, I can provide the C# code here as a comment.

**Comment by echedey-ls** [comment]
Glad to hear that @kurt-rhee !

I will have to write the tests and from a quick glance at the papers, I don't have a clue on where to start. So maybe checking against that could be an option? Maybe there's an easier path. I don't really see another reason to share them, but who knows.

If it's not a big deal for you then, please provide it. Anyways, thanks for the fast response!

**Comment by kurt-rhee** [comment]
Of course!  I would like to see if the maintainers are interested in including this feature in pvlib though before starting.  Since there is no paper that specifies the algorithm, it may fall beneath the bar for inclusion.

**Comment by mikofski** [comment]
@kurt-rhee  there is that Terabase link right in Atlassian, any way to upload that to zenodo?

**Comment by kurt-rhee** [comment]
@mikofski absolutely, I'll shoot for having a copy up there in one week

**Comment by kandersolar** [comment]
I think there is room for these kinds of models in pvlib.  @kurt-rhee is a paper for this model in the works?  I need to read that docs page in more detail, but on first glance I'm not sure its content is detailed enough to act as a reference for a pvlib implementation.

Also FYI I notice a minor inconsistency on that webpage: it says that $\mu$ and $\eta$ range from 0-100, but the equation for $\theta_c$ treats them as quantities from 0-1.

**Comment by kurt-rhee** [comment]
@kandersolar thanks for pointing out that error in the documentation, I have corrected it.

As for the paper, I haven't started it, but I wouldn't mind writing one.  I only have experience writing papers for the PVSC conferences, do you have any suggestions of where else I could submit it?  We just passed the submission window for this year's PVSC.

**Comment by kurt-rhee** [comment]
@mikofski 

I wasn't able to get into Zenodo unfortunately, looks like they are having some issues with their login right now.  I tried on firefox and chrome. Open issue on github (https://github.com/zenodo/zenodo/issues/2176)

**Comment by mikofski** [comment]
PVPMC would be a good reference - wish we could also share some validation data, maybe check with @williamhobbs or @abhisheksparikh

**Comment by williamhobbs** [comment]
I'd be glad to chat about possible validation data. I think folks can find my work email via LinkedIn, or personal email in the pvlib google group. 

Outside of NREL arrays, the NIST demo site, etc., some relatively new _public_ datasets options are:
 - DOE Solar Data Bounty Prize sites: see https://openei.org/wiki/PVDAQ, https://doi.org/10.25984/1846021 (and the upcoming DOE/NREL webinar)
 - EPRI SSRC: data here https://github.com/epri-dev/PV-Subhourly-Clipping, briefly discussed here https://github.com/pvlib/pvlib-python/discussions/1889#discussion-5743633. An EPRI report describing the dataset in more detail is pending, but I'd be glad to chat offline to fill in gaps.

**Comment by adriesse** [comment]
> PVPMC would be a good reference

This is an open, unanswered question: #1898

**Comment by kurt-rhee** [comment]
@mikofski I agree that some validation would be great.  I won't be given time this year to work on a validation study, but if anybody here now or in the future is interested in studying it I can provide the code in python, C#, or a research license to PlantPredict

**Comment by kurt-rhee** [comment]
Bringing this back up for discussion.  In the medium term future I'm going to include this in Proximal's model and I would be happy to upstream it to pvlib so that others can use it as well.  

I'm also happy to write a paper/whitepaper/blog-post/pvpmc website addition/anything since I'm not sure if the plantpredict documentation alone is enough.  

My questions are as follows:

1. Are the maintainers interested in including an algorithm like this in pvlib?
2. If so, can we agree on where the best place would be to create a reference?

**Comment by wholmgren** [comment]
The industry absolutely needs diffuse boost algorithm(s) in pvlib. I will offer my DNV time to help.

**Comment by adriesse** [comment]
This may be a silly question, but aren't these algorithms often proprietary and secret?

**Comment by williamhobbs** [comment]
> This may be a silly question, but aren't these algorithms often proprietary and secret?

I think you may be correct, but having some benchmark model(s) could be very useful. I’ve also seen patents that sound like they could include aspects of diffuse optimization, but I didn’t look close enough to be sure. 

And a reminder of the related conversation on the mailing list here, in case anyone hasn’t seen it or they forgot: https://groups.google.com/g/pvlib-python/c/uYqWRpz_cwk/m/EqY1InDCAAAJ

**Comment by kurt-rhee** [comment]
Good question from @adriesse I don't think we need to aim to recreate an existing diffuse stow algorithm from a particular company, something generic that roughly models the behavior of diffuse stowing trackers would satisfy my requirements.

Plus I would guess that diffuse stow algorithms for tracker companies need to take into account lots of non-idealities like battery state of charge, tracker turn speed, tracker motor wear, certainty of diffuse conditions, rate of change of diffuse conditions, that we could just idealize or make rough models of in the pvlib implementation.

**Comment by adriesse** [comment]
Ok, so it sounds like tracker people need to get together and discuss what the generic algorithm(s) for benchmarking should be and document that as a first step.  In [this poster from EU-PVSEC 2025](https://www.researchgate.net/publication/406821946_Instrumentation_to_Evaluate_Single-Axis_Tracker_Operation_and_Irradiance_Transposition_Calculations) we simply triggered the horizontal position based on diffuse fraction.

BTW, I'm assuming diffuse **stow**, **optimization** and **boost** are all referring to the same thing!

**Comment by cwhanse** [comment]
> something generic that roughly models the behavior of diffuse stowing trackers

I think we want to prioritize simplicity in the generic (benchmark) algorithm. What about "return to horizontal when POA at 0 degrees exceeds POA at sun-tracking angle"?

The value is to have something in pvlib, and if it's simple, it can serve as a benchmark for consistent comparison with other algorithms.

> the best place would be to create a reference

If need be, Sandia can publish a technical report, at least that creates a persistent reference.

**Comment by kurt-rhee** [comment]
I like the Sandia technical report option.

In my experience, diffuse stow algorithms in the field don't usually go to zero, they search a possibility space of different angles and choose the best possible angle.  

Might be interesting to have both algorithms (a flat diffuse stow and a scanning diffuse stow) so that we can do a comparison?

**Comment by mikofski** [comment]
Hi @kurt-rhee and @wholmgren, thank you for continuing to push this forward. A few random thoughts:
1. I agree with Kurt, most algorithms probably do not return to zero, which I believe is what PVsyst does if you check “diffuse optimization” which we’ve found to be extremely conservative. Perhaps this is one option we could offer to provide a lower bound as worse case
2. the “perfect foresight” example I shared in the Google group is also unrealistic imo and could serve as an upper bound
3. Perhaps Kevin’s & Adam’s recent generic rolling terrain backtracking paper & code can serve as an example of how to move forward? Kurt, would you publish a generic diffuse body algorithm as a reference, perhaps as Sandia technical report, pvpmc talk, or whatever?

**Comment by kurt-rhee** [comment]
I have a paper I'm working on, will probably be done in the next few days.   I can post it here first so that people can decide where the best place might be for it to land as a reference.

**Comment by kurt-rhee** [comment]
[Irradiance_Optimization.pdf](https://github.com/user-attachments/files/29059124/Irradiance_Optimization.pdf)

Here is draft of the paper.  All feedback welcome.

**Comment by williamhobbs** [comment]
@kurt-rhee, here's some quick* feedback:
- visual illustrations (tracker theta plots) of the impact of varying the values of rotation speed and hesitation would be helpful
- in table A1, case 1, it seems the tracker speed should be infinite (or very large) instead of zero?
- the link for ref 15 (the awesome google group discussion) doesn't seem to work for me
- it relies heavily on online software documentation (PVsyst a bit, Plant Predict a lot). Recreating what's in those references in the paper somehow, and/or including links to archived copies of those pages archive.org, could go a long way

*I gave it a thorough skim, and hope to review more in-depth, but figured quick feedback now is better than detailed feedback that might never show up...

**Comment by kurt-rhee** [comment]
@williamhobbs thank you!

- That is a great idea.  I'm still not quite sure if this feature is going to be accepted or not so I haven't taken the time to write the code.  If I had a better idea of whether or not this is a good fit for pvlib I will definitely include these plots in the next draft of the "paper".
- Good catch.  Originally when I was at PlantPredict I wrote this algorithm as seconds per degree instead of degrees per second.  They have since flipped that, which I think is the right decision.  I'll revise in the next draft.
- That is interesting, it works fine for me.  I'm not sure what I should do to make sure it works for everyone.  Suggestions very welcome.
- Good idea.  Yes the algorithm was originally written for PlantPredict and I've written about it in this paper with their permission.  I like the archive.org idea!

**Comment by williamhobbs** [comment]
The links that wrap over to the next line seem broken for me. I realized ref 11 is the same. Viewing the pdf in Firefox and Edge. Not that it matters too much at this stage...

<img width="1207" height="170" alt="Image" src="https://github.com/user-attachments/assets/2eef648f-093c-4c8e-900c-9e393b4d7c19" />

**Comment by wholmgren** [comment]
Thanks @kurt-rhee nice paper! From my point of view we are making this way too complicated. The two algorithms defined here are quite straightforward (yay) and I think they don't need any more peer review than what we'd see in a PR. Personally I'd rather see this paper rendered into a pvlib documentation page and we call it done.

**Comment by adriesse** [comment]
> Thanks [@kurt-rhee](https://github.com/kurt-rhee) nice paper! From my point of view we are making this way too complicated. The two algorithms defined here are quite straightforward (yay) and I think they don't need any more peer review than what we'd see in a PR. Personally I'd rather see this paper rendered into a pvlib documentation page and we call it done.

@wholmgren I would hold back a bit on the acceptance until the dust settles and the algorithm is finalized/simplified. It also makes a difference whether this is going to be called: ```singleaxis_baseline``` or ```singleaxis_plantpredict```.

**Comment by kurt-rhee** [comment]
Sounds like perhaps people prefer that the algorithm be simpler than the current proposed algorithm. I think that we could do two PR's for this task, one algorithm where the tracker goes to stow at 0 degrees when diffuse fraction goes above some number (we can let the user decide in the function arguments) which would be a simple version that people could use and inspect and also include the more complicated plantpredict version so that they can compare. 

Any thougths on this approach?

**Comment by cwhanse** [comment]
@kurt-rhee I'm in favor of the simple stow approach you describe. I would like to see a bit more mathematics in the paper, so it's easier to trace code back to the reference.
